# Data exploration and preprocessing

In [ ]:
import pandas as pd
import numpy as np
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
import tensorflow as tf
import numpy as np
from sklearn.model_selection import ParameterGrid


# Load the dataset
df = pd.read_csv('/content/Alphabets_data.csv')
df.head()

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
0      T     2     8      3       5      1     8    13      0      6      6   
1      I     5    12      3       7      2    10     5      5      4     13   
2      D     4    11      6       8      6    10     6      2      6     10   
3      N     7    11      6       6      3     5     9      4      6      4   
4      G     2     1      3       1      1     8     6      6      6      6   

   x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0      10       8      0       8      0       8  
1       3       9      2       8      4      10  
2       3       7      3       7      3       9  
3       4      10      6      10      2       8  
4       5       9      1       7      5      10

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   letter  20000 non-null  object
 1   xbox    20000 non-null  int64 
 2   ybox    20000 non-null  int64 
 3   width   20000 non-null  int64 
 4   height  20000 non-null  int64 
 5   onpix   20000 non-null  int64 
 6   xbar    20000 non-null  int64 
 7   ybar    20000 non-null  int64 
 8   x2bar   20000 non-null  int64 
 9   y2bar   20000 non-null  int64 
 10  xybar   20000 non-null  int64 
 11  x2ybar  20000 non-null  int64 
 12  xy2bar  20000 non-null  int64 
 13  xedge   20000 non-null  int64 
 14  xedgey  20000 non-null  int64 
 15  yedge   20000 non-null  int64 
 16  yedgex  20000 non-null  int64 
dtypes: int64(16), object(1)
memory usage: 2.6+ MB


In [ ]:
# Summarize key features
print("Number of samples:", len(df))
print("Number of features:", len(df.columns) - 1)  # Excluding the target column
print("Number of classes:", len(df['letter'].unique()))

Number of samples: 20000
Number of features: 16
Number of classes: 26


In [ ]:
# Split features and target variable
X = df.iloc[:,1:16]
y = df['letter']

In [ ]:
### Normalization
from sklearn.preprocessing import StandardScaler
# Scale the features
scaler = StandardScaler()
X = scaler.fit_transform(X)

In [ ]:
### label encoding
from sklearn.preprocessing import LabelEncoder
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)

In [ ]:
#### Split into training and testing
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.20, random_state=41)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((16000, 15), (4000, 15), (16000,), (4000,))

# model implementation

In [ ]:
# Initialize the ANN
model = Sequential()

In [ ]:
 #Add the input layer and the first hidden layer
model.add(Dense(units=16, activation='relu', input_dim=X_train.shape[1]))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
# Add the second hidden layer (optional)
model.add(Dense(units=8, activation='relu'))

In [ ]:
# Add the output layer
model.add(Dense(units=len(np.unique(y)), activation='softmax'))

In [ ]:
# Compile the ANN
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

In [ ]:
# Summary of the model
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 16)                  │             256 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 8)                   │             136 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 26)                  │             234 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 626 (2.45 KB)

 Trainable params: 626 (2.45 KB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
# Train the model on the training set
history = model.fit(X_train, y_train, epochs=50, batch_size=10, validation_split=0.1)

Epoch 1/50
1440/1440 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.1712 - loss: 2.8185 - val_accuracy: 0.5331 - val_loss: 1.5709
Epoch 2/50
1440/1440 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.5612 - loss: 1.4369 - val_accuracy: 0.6388 - val_loss: 1.2567
Epoch 3/50
1440/1440 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.6423 - loss: 1.1758 - val_accuracy: 0.6712 - val_loss: 1.1441
Epoch 4/50
1440/1440 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.6666 - loss: 1.0989 - val_accuracy: 0.6875 - val_loss: 1.0671
Epoch 5/50
1440/1440 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.6897 - loss: 1.0289 - val_accuracy: 0.7006 - val_loss: 1.0237
Epoch 6/50
1440/1440 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7061 - loss: 0.9735 - val_accuracy: 0.7275 - val_loss: 0.9666
Epoch 7/50
1440/1440 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7242 - loss: 0.9342 - val_accuracy: 0.7406 - val_loss: 0.9347
Epoch 8/50
1440/1440 ━━━━━━━━━━━━━━━━━━━━ 5s 1ms/step - accuracy: 0.7303 - loss: 0.8900 - 

In [ ]:
# evaluate the model
# Evaluate the model on the train set
loss, accuracy = model.evaluate(X_train, y_train)
print(f'train Accuracy: {accuracy * 100:.2f}%')

500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.8270 - loss: 0.5757
train Accuracy: 82.07%


In [ ]:
# Make predictions on the train set
yhat_train = model.predict(X_train)
y_pred_classes = yhat_train.argmax(axis=-1)

500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step


In [ ]:
from sklearn.metrics import confusion_matrix, classification_report
print(confusion_matrix(y_train, y_pred_classes ))
print(classification_report(y_train, y_pred_classes ))

[[571   0   1   0   0   0   9   1   1   3   4   2   7   0   0   5   4  13
    3   0   1   2   1   0   4   1]
 [  0 482   0  28   3   7   2   0   0   0   1   0   0   0   4   4   2  28
   18   6   0  11   1   4  18   0]
 [  7   0 485   0  29   1  44   3   0   0   3   0   0   0  11   0   6   0
    2   7   8   0   0   4   0   0]
 [  0  75   0 496   0   2   0   9   0   6   0   0   3   6   7   4   4  26
    1   0   0   0   0   3   2   4]
 [  1   0   1   7 449   4  24   0   6   0   7   3   2   0   0   0  16   1
   22  25   0   2   0  19   1  16]
 [  0  37   3   9   1 497   1   0   5   3   2   0   0   1   4  14   0   0
    6  12   0   2   3   2  15   1]
 [  1  19  17   0   3   2 417   4   0   2   3   0   2   0  14   3  58   9
   21   2   2  25   0   2   8   0]
 [  3  11   0  24   4   6   9 388   0   2  19   0   1   5  18   0   6  62
    4   0  17   4   0   0   0   0]
 [  0   4   0   1   1   7   0   0 532  13   0   2   0   0   0   5   1   0
   30   0   0   0   0   4   1   8]
 [  4   6   0   5  

In [ ]:
# Make predictions on the test set
yhat_test = model.predict(X_test)
y_pred_classes1 = yhat_test.argmax(axis=-1)

125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step


In [ ]:
from sklearn.metrics import confusion_matrix, classification_report
print(confusion_matrix(y_test, y_pred_classes1 ))
print(classification_report(y_test, y_pred_classes1 ))

[[145   0   0   0   0   0   2   0   0   0   2   0   0   0   0   0   2   1
    1   0   0   1   0   0   2   0]
 [  0 109   0   8   1   1   0   0   0   0   1   0   1   0   2   1   0   8
    4   2   0   4   1   0   4   0]
 [  1   0 100   0   5   2   4   1   1   0   1   0   0   0   7   0   0   0
    0   0   2   0   0   0   2   0]
 [  0  24   0 111   0   0   0   4   0   4   0   0   1   3   1   1   1   5
    0   0   0   0   0   2   0   0]
 [  0   0   0   1 123   3   5   0   2   1   4   0   1   0   1   0   5   4
    3   2   0   0   0   2   0   5]
 [  1  12   0   5   0 124   0   0   1   0   0   0   1   0   0   3   0   0
    3   3   0   0   0   1   3   0]
 [  1   3   2   1   1   0 111   1   0   1   0   0   0   0   1   1  14   3
    9   1   0   5   1   0   3   0]
 [  2   2   0   6   1   1   0  95   0   0   2   0   1   4  13   0   3  16
    2   0   2   1   0   0   0   0]
 [  0   1   0   1   1   2   0   0 130   0   0   0   0   0   0   1   0   0
    7   1   0   0   0   2   0   0]
 [  1   2   0   2  

the model performs consistently well with an overall accuracy of aroud 80% on both the training and test sets.

# Hyperparameter tuning

In [ ]:
def create_model(hidden_layers=1, units=15, activation='relu', learning_rate=0.001):  # Model Creation Function
    model = Sequential()
    model.add(Dense(units=units, activation=activation, input_dim=X_train.shape[1]))

    for _ in range(hidden_layers - 1):
        model.add(Dense(units=units, activation=activation))

    model.add(Dense(units=len(np.unique(y)), activation='softmax'))

    optimizer = tf.keras.optimizers.Adam(learning_rate=learning_rate)
    model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    return model

In [ ]:
def evaluate_model(hidden_layers, units, activation, learning_rate):        #Model Evaluation Function
    model = create_model(hidden_layers=hidden_layers, units=units, activation=activation, learning_rate=learning_rate)
    model.fit(X_train, y_train, epochs=10, batch_size=10, verbose=0)
    _, accuracy = model.evaluate(X_test, y_test, verbose=0)
    return accuracy

In [ ]:
#Perform Hyperparameter Tuning
param_grid = {
    'hidden_layers': [1, 2, 3],
    'units': [8, 16, 32],
    'activation': ['relu', 'tanh'],
    'learning_rate': [0.001, 0.01]
}

best_score = 0
best_params = {}

for params in ParameterGrid(param_grid):
    score = evaluate_model(**params)
    print(f"Params: {params} - Score: {score}")
    if score > best_score:
        best_score = score
        best_params = params

print(f"Best Score: {best_score}")
print(f"Best Params: {best_params}")

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Params: {'activation': 'relu', 'hidden_layers': 1, 'learning_rate': 0.001, 'units': 8} - Score: 0.7182499766349792
Params: {'activation': 'relu', 'hidden_layers': 1, 'learning_rate': 0.001, 'units': 16} - Score: 0.7737500071525574
Params: {'activation': 'relu', 'hidden_layers': 1, 'learning_rate': 0.001, 'units': 32} - Score: 0.8399999737739563
Params: {'activation': 'relu', 'hidden_layers': 1, 'learning_rate': 0.01, 'units': 8} - Score: 0.7245000004768372
Params: {'activation': 'relu', 'hidden_layers': 1, 'learning_rate': 0.01, 'units': 16} - Score: 0.793749988079071
Params: {'activation': 'relu', 'hidden_layers': 1, 'learning_rate': 0.01, 'units': 32} - Score: 0.8579999804496765
Params: {'activation': 'relu', 'hidden_layers': 2, 'learning_rate': 0.001, 'units': 8} - Score: 0.703249990940094
Params: {'activation': 'relu', 'hidden_layers': 2, 'learning_rate': 0.001, 'units': 16} - Score: 0.796750009059906
Params: {'activation': 'relu', 'hidden_layers': 2, 'learning_rate': 0.001, 'units

In [ ]:
# Best parameters from the tuning process
best_hidden_layers = best_params['hidden_layers']
best_units = best_params['units']
best_activation = best_params['activation']
best_learning_rate = best_params['learning_rate']

In [ ]:
# Create the final model with the best parameters
final_model = create_model(
    hidden_layers=best_hidden_layers,
    units=best_units,
    activation=best_activation,
    learning_rate=best_learning_rate
)

In [ ]:
# Train the final model on the full training data
final_model.fit(X_train, y_train, epochs=10, batch_size=10, verbose=1)

Epoch 1/10
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.4485 - loss: 2.1038
Epoch 2/10
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 6s 2ms/step - accuracy: 0.7431 - loss: 0.9545
Epoch 3/10
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.7913 - loss: 0.7392
Epoch 4/10
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8177 - loss: 0.6340
Epoch 5/10
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8468 - loss: 0.5354
Epoch 6/10
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.8649 - loss: 0.4818
Epoch 7/10
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8724 - loss: 0.4312
Epoch 8/10
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 6s 2ms/step - accuracy: 0.8814 - loss: 0.4025
Epoch 9/10
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8935 - loss: 0.3629
Epoch 10/10
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8934 - loss: 0.3458


In [ ]:
#Import necessary libraries
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report

In [ ]:
 #Make predictions on the train set
yhat_train = final_model.predict(X_train)
y_pred_classes = yhat_train.argmax(axis=-1)

500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step


In [ ]:
# Make predictions on the test set
yhat_test = final_model.predict(X_test)
y_pred_classes1 = yhat_test.argmax(axis=-1)

125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step


In [ ]:
from sklearn.metrics import confusion_matrix, classification_report
print(confusion_matrix(y_test, y_pred_classes1 ))
print(classification_report(y_test, y_pred_classes1 ))

[[153   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0
    0   0   1   0   0   0   1   0]
 [  0 125   0   4   1   1   0   0   0   1   1   0   0   1   0   0   0   8
    3   1   1   0   0   0   0   0]
 [  0   0 107   1   5   0   0   0   0   0   2   0   0   0   6   0   0   0
    0   1   1   0   3   0   0   0]
 [  1  14   0 123   0   1   0   5   0   1   1   0   2   1   2   0   0   5
    0   0   1   0   0   0   0   0]
 [  1   0   2   0 142   3   4   0   0   0   2   0   0   0   0   0   1   1
    1   0   0   1   0   0   0   4]
 [  0   1   0   1   0 146   0   0   1   0   0   0   0   0   0   2   0   0
    3   2   0   0   0   1   0   0]
 [  1   0   3   2   2   0 130   0   0   0   2   0   0   1   1   1   1   2
    6   0   4   2   0   1   0   0]
 [  0   2   0   1   0   2   0 120   0   2   3   0   0   6   2   0   1   9
    0   0   0   2   0   0   1   0]
 [  0   0   0   3   0   0   0   0 131   4   0   0   0   0   0   1   1   0
    3   0   0   0   0   3   0   0]
 [  1   0   0   1  

training set metrics precision,recall,and f1-score:
the ,odel demonstrates high precision and recall across most classes,with some minor variations.
for instance, class 1 has a precision of 0.82 and a recall of 0.93, leading to an f1-score of 0.87, which is lower compared to other classes but still acceptable.class 24 shows excellent performance with a precision of 0.97, recall of 0.91, and an f1-score of 0.94.
testing set metrics
precision,recall, and f1-score:
the model maintains good precision and recall, though slightly lower than on the training set.
for example, class 1 has a precision of 0.80 and a recall of 0.85, resulting in an f1-score of 0.82.


Tuned Model
Training Accuracy: 91%
Testing Accuracy: 89%
Default Model
Training Accuracy: 81.24%
Testing Accuracy: 80%
Overall Improvement: The tuned model outperforms the default model in both training and testing accuracy. The tuned model has higher precision, recall, and F1-scores across most classes.
Consistency: The performance improvements are consistent across various metrics, indicating a well-tuned model.